runtime/near-vm-runner/src/cache.rs

```diff
@@ -141,9 +142,85 @@ pub(crate) fn compile_module_cached_wasmer(
+    let key = get_key(wasm_code_hash, wasm_code, VMKind::Wasmer0, config);
     #[cfg(not(feature = "no_cache"))]
     return memcache_compile_module_cached_wasmer(key, wasm_code, config, cache);
     #[cfg(feature = "no_cache")]
     return compile_module_cached_wasmer_impl(key, wasm_code, config, cache);
 }
+
+#[cfg(feature = "wasmer1_vm")]
+pub(crate) fn compile_module_cached_wasmer1(
+    wasm_code_hash: &[u8],
+    wasm_code: &[u8],
+    config: &VMConfig,
+    cache: Option<&dyn CompiledContractCache>,
+    store: &wasmer::Store,
+) -> Result<wasmer::Module, VMError> {
+    let key = get_key(wasm_code_hash, wasm_code, VMKind::Wasmer1, config);
+    return compile_module_cached_wasmer1_impl(key, wasm_code, cache, store);
+}
+
+#[cfg(feature = "wasmer1_vm")]
+fn compile_module_wasmer1(
+    prepared_code: &[u8],
+    store: &wasmer::Store,
+) -> Result<wasmer::Module, VMError> {
+    wasmer::Module::new(&store, prepared_code).map_err(|err| err.into_vm_error())
+}
+
+#[cfg(feature = "wasmer1_vm")]
+pub(crate) fn compile_and_serialize_wasmer1(
+    wasm_code: &[u8],
+    key: &CryptoHash,
+    cache: &dyn CompiledContractCache,
+    store: &wasmer::Store,
+) -> Result<wasmer::Module, VMError> {
+    let module =
+        compile_module_wasmer1(wasm_code, store).map_err(|e| cache_error(e, &key, cache))?;
+    let code = module
+        .serialize()
+        .map_err(|_e| VMError::CacheError(SerializationError { hash: (key.0).0 }))?;
+    let serialized = CacheRecord::Code(code).try_to_vec().unwrap();
+    cache.put(key.as_ref(), &serialized).map_err(|_e| VMError::CacheError(WriteError))?;
+    Ok(module)
+}
+
+#[cfg(feature = "wasmer1_vm")]
+fn deserialize_wasmer1(
+    serialized: &[u8],
+    store: &wasmer::Store,
+) -> Result<Result<wasmer::Module, VMError>, CacheError> {
+    let record = CacheRecord::try_from_slice(serialized).map_err(|_e| DeserializationError)?;
+    let serialized_module = match record {
+        CacheRecord::Error(err) => return Ok(Err(err)),
+        CacheRecord::Code(code) => code,
+    };
+    unsafe {
+        Ok(Ok(wasmer::Module::deserialize(store, serialized_module.as_slice())
+            .map_err(|_e| CacheError::DeserializationError)?))
+    }
+}
+
+#[cfg(feature = "wasmer1_vm")]
+fn compile_module_cached_wasmer1_impl(
+    key: CryptoHash,
+    wasm_code: &[u8],
+    cache: Option<&dyn CompiledContractCache>,
+    store: &wasmer::Store,
+) -> Result<wasmer::Module, VMError> {
+    if cache.is_none() {
+        return compile_module_wasmer1(wasm_code, store);
+    }
+
+    let cache = cache.unwrap();
+    match cache.get(&(key.0).0) {
+        Ok(serialized) => match serialized {
+            Some(serialized) => {
+                deserialize_wasmer1(serialized.as_slice(), store).map_err(VMError::CacheError)?
+            }
+            None => compile_and_serialize_wasmer1(wasm_code, &key, cache, store),
+        },
+        Err(_) => Err(VMError::CacheError(ReadError)),
+    }
+}
```

Comment:
This can be much easier if we moved VM specific caching into separate mods. E.g. you have mod (as a folder) `wasmer_0_17` and mod `wasmer_1`. They have the same interface and you only need to import them once. Instead of having to mark each method with `#[cfg(feature = "wasmer1_vm")]`

Reply:
Good idea